File: 1914-cyclically-rotating-a-grid/1914-cyclically-rotating-a-grid.java
class Solution {
      public int[][] rotateGrid(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;
        
        for (int i = 0; i < Math.min(m/2, n/2); i++) {
            int div = (m - i*2)*2 + (n - i*2 -2)*2;
            rotate(grid, i, i, k%div);
        }
        return grid;
    }
    
    public void rotate(int[][] matrix, int row, int col, int k) {
        Queue<Integer> list = new LinkedList<>();
        int m = matrix.length;
        int n = matrix[0].length;
        for (int i = col; i <= n-col-1; i++) list.add(matrix[row][i]);
        for (int i = row+1; i <= m-row-1-1; i++) list.add(matrix[i][n-col-1]);
        for (int i = n-col-1; i >= col; i--) list.add(matrix[m-row-1][i]);
        for (int i = m-row-1-1; i >= row+1; i--) list.add(matrix[i][col]);
        
        // cyclic rotate k times.
        while (k > 0) {
            int tmp = list.poll();
            list.add(tmp);
            k--;
        }
        
        for (int i = col; i <= n-col-1; i++) matrix[row][i] = list.poll();
        for (int i = row+1; i <= m-row-1-1; i++) matrix[i][n-col-1] = list.poll();
        for (int i = n-col-1; i >= col; i--) matrix[m-row-1][i] = list.poll();
        for (int i = m-row-1-1; i >= row+1; i--) matrix[i][col] = list.poll();
        
    }
}

File: 1914-cyclically-rotating-a-grid/1914-cyclically-rotating-a-grid.cpp
class Solution {
public:
    vector<vector<int>> rotateGrid(vector<vector<int>>& matrix, int R) {
        int M=matrix.size();
        int N=matrix[0].size();
        int numRings = min(M,N)/2;
        for(int i = 0; i < numRings; i++) {
            int numRotations = R%(2*(M + N - 4*i) - 4);
            for(int rotation = 0; rotation < numRotations; rotation++) {
                // Now rotate the ring
                // Rotate top row
                for(int j = i; j < N-i-1; j++) {
                    int tmp = matrix[i][j];
                    matrix[i][j] = matrix[i][j+1];
                    matrix[i][j+1] = tmp;
                }
                // Rotate right column
                for(int j = i; j < M-i-1; j++) {
                    int tmp = matrix[j][N-i-1];
                    matrix[j][N-i-1] = matrix[j+1][N-i-1];
                    matrix[j+1][N-i-1] = tmp;
                }
                // Rotate bottom row
                for(int j = N-i-1; j > i; j--) {
                    int tmp = matrix[M-i-1][j];
                    matrix[M-i-1][j] = matrix[M-i-1][j-1];
                    matrix[M-i-1][j-1] = tmp;
                }
                // Rotate left column
                for(int j = M-i-1; j > i+1; j--) {
                    int tmp = matrix[j][i];
                    matrix[j][i] = matrix[j-1][i];
                    matrix[j-1][i] = tmp;
                }
            }
        }
        return matrix;
    }
};